### Exercise 1: Get the rock sample data into Visual Studio Code

<font size="3"> **Unit 3 of 11** </font>

<font size="2">Step by step process of **Exercise 1** can be found here: https://learn.microsoft.com/en-us/training/modules/plan-moon-mission-using-python-pandas/3-setup-import-data</font>

In [1]:
import pandas as pd
rock_samples = pd.read_csv('data/rocksamples.csv') 

rock_samples.head()

rock_samples.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2229 entries, 0 to 2228
Data columns (total 6 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   ID            2229 non-null   int64  
 1   Mission       2229 non-null   object 
 2   Type          2229 non-null   object 
 3   Subtype       2226 non-null   object 
 4   Weight (g)    2229 non-null   float64
 5   Pristine (%)  2229 non-null   float64
dtypes: float64(2), int64(1), object(3)
memory usage: 104.6+ KB


### Exercise 2: Determine the question to ask to inform data cleansing

<font size="3"> **Unit 4 of 11** </font>

<font size="2">Step by step process of **Exercise 2** can be found here: https://learn.microsoft.com/en-us/training/modules/plan-moon-mission-using-python-pandas/4-setup-cleanse.data </font>

#### Exercise 2.1

In [3]:
import pandas as pd
rock_samples = pd.read_csv('data/rocksamples.csv') 

rock_samples['Weight (g)'] = rock_samples['Weight (g)'].apply(lambda x : x * 0.001)
rock_samples.rename(columns={'Weight (g)' : 'Weight (kg)'}, inplace = True)
rock_samples.head()


ID   Mission    Type   Subtype  Weight (kg)  Pristine (%)
0  10001  Apollo11    Soil  Unsieved       0.1258         88.36
1  10002  Apollo11    Soil  Unsieved       5.6290         93.73
2  10003  Apollo11  Basalt  Ilmenite       0.2130         65.56
3  10004  Apollo11    Core  Unsieved       0.0448         71.76
4  10005  Apollo11    Core  Unsieved       0.0534         40.31

#### Exercise 2.2

In [5]:
import pandas as pd
rock_samples = pd.read_csv('data/rocksamples.csv')

missions = pd.DataFrame()
missions['Mission'] = rock_samples['Mission'].unique()
missions.head()

Mission
0  Apollo11
1  Apollo12
2  Apollo14
3  Apollo15
4  Apollo16

#### Exercise 2.3

In [10]:
import pandas as pd
rock_samples = pd.read_csv('data/rocksamples.csv')

rock_samples['Weight (g)'] = rock_samples['Weight (g)'].apply(lambda x : x * 0.001)
rock_samples.rename(columns = {'Weight (g)' : 'Weight (kg)'}, inplace = True)
rock_samples.head()

missions = pd.DataFrame()
missions['Mission'] = rock_samples['Mission'].unique()
missions.head()

sample_total_weight = rock_samples.groupby('Mission')['Weight (kg)'].sum()
missions = pd.merge(missions, sample_total_weight, on = 'Mission')
missions.rename(columns = {'Weight (kg)' : 'Sample Weight (kg)'}, inplace  = True)
missions

Mission  Sample Weight (kg)
0  Apollo11            21.55424
1  Apollo12            34.34238
2  Apollo14            41.83363
3  Apollo15            75.39910
4  Apollo16            92.46262
5  Apollo17           109.44402

#### Exercise 2.4

In [13]:
import pandas as pd
rock_samples = pd.read_csv('data/rocksamples.csv')

rock_samples ['Weight (g)'] = rock_samples['Weight (g)'].apply(lambda x : x * 0.001)
rock_samples.rename(columns={'Weight (g)' : 'Weight (kg)'}, inplace = True)
rock_samples.head()

missions = pd.DataFrame()
missions ['Mission'] = rock_samples ['Mission'].unique()
missions.head()

sample_total_weight = rock_samples.groupby('Mission')['Weight (kg)'].sum()
sample_total_weight

Mission
Apollo11     21.55424
Apollo12     34.34238
Apollo14     41.83363
Apollo15     75.39910
Apollo16     92.46262
Apollo17    109.44402
Name: Weight (kg), dtype: float64

#### Exercise 2.5

In [29]:
import pandas as pd
rock_samples = pd.read_csv('data/rocksamples.csv')

rock_samples ['Weight (g)'] = rock_samples ['Weight (g)'].apply(lambda x : x * 0.001)
rock_samples.rename(columns={'Weight (g)' : 'Weight (kg)'}, inplace = True)
rock_samples.head()

missions = pd.DataFrame()
missions ['Mission'] = rock_samples['Mission'].unique()
missions.head()

sample_total_weight = rock_samples.groupby('Mission')['Weight (kg)'].sum()
missions = pd.merge(missions, sample_total_weight, on = 'Mission')
missions.rename(columns = {'Weight (kg)' : 'Sample Weight (kg)'}, inplace  = True)
missions['Weight Difference'] = missions['Sample Weight (kg)'].diff()
missions

Mission  Sample Weight (kg)  Weight Difference
0  Apollo11            21.55424                NaN
1  Apollo12            34.34238           12.78814
2  Apollo14            41.83363            7.49125
3  Apollo15            75.39910           33.56547
4  Apollo16            92.46262           17.06352
5  Apollo17           109.44402           16.98140

#### Exercise 2.6

In [1]:
import pandas as pd
rock_samples = pd.read_csv('data/rocksamples.csv')

rock_samples ['Weight (g)'] = rock_samples ['Weight (g)'].apply(lambda x : x * 0.001)
rock_samples.rename(columns={'Weight (g)' : 'Weight (kg)'}, inplace = True)
rock_samples.head()

missions = pd.DataFrame()
missions ['Mission'] = rock_samples['Mission'].unique()
missions.head()

sample_total_weight = rock_samples.groupby('Mission')['Weight (kg)'].sum()
missions = pd.merge(missions, sample_total_weight, on = 'Mission')
missions.rename(columns = {'Weight (kg)' : 'Sample Weight (kg)'}, inplace  = True)
missions['Weight Difference'] = missions['Sample Weight (kg)'].diff()
missions['Weight Difference'] = missions['Weight Difference'].fillna(value = 0)
missions

Mission  Sample Weight (kg)  Weight Difference
0  Apollo11            21.55424            0.00000
1  Apollo12            34.34238           12.78814
2  Apollo14            41.83363            7.49125
3  Apollo15            75.39910           33.56547
4  Apollo16            92.46262           17.06352
5  Apollo17           109.44402           16.98140

### Exercise 3: Add rocket weight data to the mission analysis

<font size="3"> **Unit 5 of 11** </font>

<font size="2">Step by step process of **Exercise 3** can be found here: https://learn.microsoft.com/en-us/training/modules/plan-moon-mission-using-python-pandas/5-add-modules </font>

------

<font size="3"> **References/Vocabs:** </font>
<font size="2">
- *__Command module:__* The module that astronauts live in. When two astronauts are down on the surface of the Moon, the third astronaut stays in the command module. This module is returned to Earth.
- *__Lunar module:__* The module that detaches from the command module after it has reached orbit around the Moon. This module lands on the surface of the Moon and can carry two astronauts. When the lunar module returns from the surface to the command module, it leaves part of the base (the landing gear) on the surface of the Moon. </font>

#### Exercise 3.1

In [2]:
import pandas as pd
rock_samples = pd.read_csv('data/rocksamples.csv')

rock_samples ['Weight (g)'] = rock_samples ['Weight (g)'].apply(lambda x : x * 0.001)
rock_samples.rename(columns={'Weight (g)' : 'Weight (kg)'}, inplace = True)
rock_samples.head()

missions = pd.DataFrame()
missions ['Mission'] = rock_samples['Mission'].unique()
missions.head()

sample_total_weight = rock_samples.groupby('Mission')['Weight (kg)'].sum()
missions = pd.merge(missions, sample_total_weight, on = 'Mission')
missions.rename(columns = {'Weight (kg)' : 'Sample Weight (kg)'}, inplace  = True)
missions['Weight Difference'] = missions['Sample Weight (kg)'].diff()
missions['Weight Difference'] = missions['Weight Difference'].fillna(value = 0)

missions['Lunar Module (LM)'] = ['Eagle (LM-5)', 'Intrepid (LM-6)', 'Antares (LM-8)', 'Falcon (LM-10)', 'Orion (LM-11)', 'Challenger (LM-12)']
missions['LM Mass (kg)'] = [15103, 15235, 15264, 16430, 16445, 16456]
missions['LM Mass Difference'] = missions['LM Mass (kg)'].diff()
missions['LM Mass Difference'] = missions['LM Mass (kg)'].fillna(value = 0)

missions['Command Module (CM)'] =['Columbia (CSM-107)', 'Yankee Clipper (CM-108)', 'Kitty Hawk (CM-110)', 'Endeavor (CM-112)', 'Casper (CM-113)', 'America (CM-114)']
missions['CM Mass (kg)'] = [5560, 5609, 5758, 5875, 5840, 5960]
missions['CM Mass Difference (kg)'] = missions['CM Mass (kg)'].diff()
missions['CM Mass Difference (kg)'] = missions['CM Mass (kg)'].fillna(value = 0)

missions

Mission  Sample Weight (kg)  Weight Difference   Lunar Module (LM)  \
0  Apollo11            21.55424            0.00000        Eagle (LM-5)   
1  Apollo12            34.34238           12.78814     Intrepid (LM-6)   
2  Apollo14            41.83363            7.49125      Antares (LM-8)   
3  Apollo15            75.39910           33.56547      Falcon (LM-10)   
4  Apollo16            92.46262           17.06352       Orion (LM-11)   
5  Apollo17           109.44402           16.98140  Challenger (LM-12)   

   LM Mass (kg)  LM Mass Difference      Command Module (CM)  CM Mass (kg)  
0         15103               15103       Columbia (CSM-107)           0.0  
1         15235               15235  Yankee Clipper (CM-108)          49.0  
2         15264               15264      Kitty Hawk (CM-110)         149.0  
3         16430               16430        Endeavor (CM-112)         117.0  
4         16445               16445          Casper (CM-113)         -35.0  
5         16456               16456         America (CM-114)         120.0

#### Exercise 3.2

In [21]:
import pandas as pd
rock_samples = pd.read_csv('data/rocksamples.csv')

rock_samples['Weight (g)'] = rock_samples['Weight (g)'].apply(lambda x : x * 0.001)
rock_samples.rename(columns = {'Weight (g)' : 'Weight (kg)'}, inplace = True)
rock_samples.head()

missions = pd.DataFrame()
missions['Mission'] = rock_samples['Mission'].unique()
missions.head()

sample_total_weight = rock_samples.groupby('Mission')['Weight (kg)'].sum()
missions = pd.merge(missions, sample_total_weight, on = 'Mission')
missions.rename(columns = {'Weight (kg)' : 'Sample Weight (kg)'}, inplace = True)
missions['Weight Difference (kg)'] = missions['Sample Weight (kg)'].diff()
missions['Weight Difference (kg)'] = missions['Weight Difference (kg)'].fillna(value = 0)

missions['Lunar Module (LM)'] = ['Eagle (LM-5)', 'Intrepid (LM-6)', 'Antares (LM-8)', 'Falcon (LM-10)', 'Orion (LM-11)', 'Challenger (LM-12)']
missions['LM Mass (kg)'] = [15103, 15235, 15264, 16430, 16445, 16456]
missions['LM Mass Difference'] = missions['LM Mass (kg)'].diff()
missions['LM Mass Difference'] = missions['LM Mass Difference'].fillna(value = 0)

missions['Command Module (CM)'] =['Columbia (CSM-107)', 'Yankee Clipper (CM-108)', 'Kitty Hawk (CM-110)', 'Endeavor (CM-112)', 'Casper (CM-113)', 'America (CM-114)']
missions['CM Mass (kg)'] = [5560, 5609, 5758, 5875, 5840, 5960]
missions['CM Mass Difference'] = missions['CM Mass (kg)'].diff()
missions['CM Mass Difference'] = missions['CM Mass Difference'].fillna(value = 0)

missions['Total Weight (kg)'] = missions['LM Mass (kg)'] + missions['CM Mass (kg)']
missions['Total Weight Difference (kg)'] = missions['LM Mass Difference'] + missions['CM Mass Difference']

missions

Mission  Sample Weight (kg)  Weight Difference (kg)   Lunar Module (LM)  \
0  Apollo11            21.55424                 0.00000        Eagle (LM-5)   
1  Apollo12            34.34238                12.78814     Intrepid (LM-6)   
2  Apollo14            41.83363                 7.49125      Antares (LM-8)   
3  Apollo15            75.39910                33.56547      Falcon (LM-10)   
4  Apollo16            92.46262                17.06352       Orion (LM-11)   
5  Apollo17           109.44402                16.98140  Challenger (LM-12)   

   LM Mass (kg)  LM Mass Difference      Command Module (CM)  CM Mass (kg)  \
0         15103                 0.0       Columbia (CSM-107)          5560   
1         15235               132.0  Yankee Clipper (CM-108)          5609   
2         15264                29.0      Kitty Hawk (CM-110)          5758   
3         16430              1166.0        Endeavor (CM-112)          5875   
4         16445                15.0          Casper (CM-113)          5840   
5         16456                11.0         America (CM-114)          5960   

   CM Mass Difference  Total Weight (kg)  Total Weight Difference (kg)  
0                 0.0              20663                           0.0  
1                49.0              20844                         181.0  
2               149.0              21022                         178.0  
3               117.0              22305                        1283.0  
4               -35.0              22285                         -20.0  
5               120.0              22416                         131.0

### Exercise 4: Understand the data in the missions DataFrame

<font size="3"> **Unit 6 of 11** </font>

<font size="2">Step by step process of **Exercise 4** can be found here: https://learn.microsoft.com/en-us/training/modules/plan-moon-mission-using-python-pandas/6-understand-modules </font>

------

<font size="3"> **References/Vocabs:** </font>

<font size="2"> A *__payload__* is basically the total amount of weight that a rocket can get up through our atmosphere and into space. So the likelihood that the payload number is more accurate than the exact weights of each module is high, because deciding the payload will likely affect each of the other design decisions.

We know that the *__Saturn V payload was 43,500 kg__*, and the weights of the modules varied from mission to mission. So, to determine the ratios that will allow us to make predictions about the Artemis missions, we can use:

- Saturn V payload
- Mission sample weight
- Mission module weight

We're calling the two modules *__crewed area__* in the DataFrame because those are the parts of the spacecraft where the crew can be, and likely where the samples would also reside.
</font>

#### Exercise 4.1

In [1]:
import pandas as pd
rock_samples = pd.read_csv('data/rocksamples.csv')

rock_samples['Weight (g)'] = rock_samples['Weight (g)'].apply(lambda x : x * 0.001)
rock_samples.rename(columns = {'Weight (g)' : 'Weight (kg)'}, inplace = True)
rock_samples.head()

missions = pd.DataFrame()
missions['Mission'] = rock_samples['Mission'].unique()
missions.head()

sample_total_weight = rock_samples.groupby('Mission')['Weight (kg)'].sum()
missions = pd.merge(missions, sample_total_weight, on = 'Mission')
missions.rename(columns = {'Weight (kg)' : 'Sample Weight (kg)'}, inplace = True)
missions['Weight Difference (kg)'] = missions['Sample Weight (kg)'].diff()
missions['Weight Difference (kg)'] = missions['Weight Difference (kg)'].fillna(value = 0)

missions['Lunar Module (LM)'] = ['Eagle (LM-5)', 'Intrepid (LM-6)', 'Antares (LM-8)', 'Falcon (LM-10)', 'Orion (LM-11)', 'Challenger (LM-12)']
missions['LM Mass (kg)'] = [15103, 15235, 15264, 16430, 16445, 16456]
missions['LM Mass Difference'] = missions['LM Mass (kg)'].diff()
missions['LM Mass Difference'] = missions['LM Mass Difference'].fillna(value = 0)

missions['Command Module (CM)'] =['Columbia (CSM-107)', 'Yankee Clipper (CM-108)', 'Kitty Hawk (CM-110)', 'Endeavor (CM-112)', 'Casper (CM-113)', 'America (CM-114)']
missions['CM Mass (kg)'] = [5560, 5609, 5758, 5875, 5840, 5960]
missions['CM Mass Difference'] = missions['CM Mass (kg)'].diff()
missions['CM Mass Difference'] = missions['CM Mass Difference'].fillna(value = 0)

missions['Total Weight (kg)'] = missions['LM Mass (kg)'] + missions ['CM Mass (kg)']
missions['Total Difference (kg)'] = missions['LM Mass Difference'] + missions['CM Mass Difference']

# Sample-to-Weight Ratio
saturnVPayload = 43500 #(kg)
missions['Crewed Area: Payload'] = missions['Total Weight (kg)']/saturnVPayload
missions['Sample: Crewed Area'] = missions['Sample Weight (kg)']/missions['Total Weight (kg)']
missions['Sample: Payload'] = missions['Sample Weight (kg)']/saturnVPayload 

missions

Mission  Sample Weight (kg)  Weight Difference (kg)   Lunar Module (LM)  \
0  Apollo11            21.55424                 0.00000        Eagle (LM-5)   
1  Apollo12            34.34238                12.78814     Intrepid (LM-6)   
2  Apollo14            41.83363                 7.49125      Antares (LM-8)   
3  Apollo15            75.39910                33.56547      Falcon (LM-10)   
4  Apollo16            92.46262                17.06352       Orion (LM-11)   
5  Apollo17           109.44402                16.98140  Challenger (LM-12)   

   LM Mass (kg)  LM Mass Difference      Command Module (CM)  CM Mass (kg)  \
0         15103                 0.0       Columbia (CSM-107)          5560   
1         15235               132.0  Yankee Clipper (CM-108)          5609   
2         15264                29.0      Kitty Hawk (CM-110)          5758   
3         16430              1166.0        Endeavor (CM-112)          5875   
4         16445                15.0          Casper (CM-113)          5840   
5         16456                11.0         America (CM-114)          5960   

   CM Mass Difference  Total Weight (kg)  Total Difference (kg)  \
0                 0.0              20663                    0.0   
1                49.0              20844                  181.0   
2               149.0              21022                  178.0   
3               117.0              22305                 1283.0   
4               -35.0              22285                  -20.0   
5               120.0              22416                  131.0   

   Crewed Area: Payload  Sample: Crewed Area  Sample: Payload  
0              0.475011             0.001043         0.000495  
1              0.479172             0.001648         0.000789  
2              0.483264             0.001990         0.000962  
3              0.512759             0.003380         0.001733  
4              0.512299             0.004149         0.002126  
5              0.515310             0.004882         0.002516

#### Exercise 4.2

In [13]:
import pandas as pd
rock_samples = pd.read_csv('data/rocksamples.csv')

rock_samples['Weight (g)'] = rock_samples['Weight (g)'].apply(lambda x : x * 0.001)
rock_samples.rename(columns = {'Weight (g)' : 'Weight (kg)'}, inplace = True)
rock_samples.head()

missions = pd.DataFrame()
missions['Mission'] = rock_samples['Mission'].unique()
missions.head()

sample_total_weight = rock_samples.groupby('Mission')['Weight (kg)'].sum()
missions = pd.merge(missions, sample_total_weight, on = 'Mission')
missions.rename(columns = {'Weight (kg)' : 'Sample Weight (kg)'}, inplace = True)
missions['Weight Difference (kg)'] = missions['Sample Weight (kg)'].diff()
missions['Weight Difference (kg)'] = missions['Weight Difference (kg)'].fillna(value = 0)

missions['Lunar Module (LM)'] = ['Eagle (LM-5)', 'Intrepid (LM-6)', 'Antares (LM-8)', 'Falcon (LM-10)', 'Orion (LM-11)', 'Challenger (LM-12)']
missions['LM Mass (kg)'] = [15103, 15235, 15264, 16430, 16445, 16456]
missions['LM Mass Difference'] = missions['LM Mass (kg)'].diff()
missions['LM Mass Difference'] = missions['LM Mass Difference'].fillna(value = 0)

missions['Command Module (CM)'] =['Columbia (CSM-107)', 'Yankee Clipper (CM-108)', 'Kitty Hawk (CM-110)', 'Endeavor (CM-112)', 'Casper (CM-113)', 'America (CM-114)']
missions['CM Mass (kg)'] = [5560, 5609, 5758, 5875, 5840, 5960]
missions['CM Mass Difference'] = missions['CM Mass (kg)'].diff()
missions['CM Mass Difference'] = missions['CM Mass Difference'].fillna(value = 0)

missions['Total Weight (kg)'] = missions['LM Mass (kg)'] + missions['CM Mass (kg)']
missions['Total Difference (kg)'] = missions['LM Mass Difference'] + missions['CM Mass Difference']

# Sample-to-weight ratio:
saturnVPayload = 43500 #(kg)
missions['Crewed Area: Payload'] = missions['Total Weight (kg)']/saturnVPayload
missions['Sample: Crewed Area'] = missions['Sample Weight (kg)']/missions['Total Weight (kg)']
missions ['Sample: Payload'] = missions['Sample Weight (kg)']/saturnVPayload

missions

# Take the average of all those ratios across all the missions:
crewedarea_payload_average = missions['Crewed Area: Payload'].mean()
sample_crewedarea_average = missions['Sample: Crewed Area'].mean()
sample_payload_average = missions['Sample: Payload'].mean()

print(crewedarea_payload_average)
print(sample_crewedarea_average)
print(sample_payload_average)

0.4963026819923371
0.002848764392685611
0.0014369195019157087


Mission  Sample Weight (kg)  Weight Difference (kg)   Lunar Module (LM)  \
0  Apollo11            21.55424                 0.00000        Eagle (LM-5)   
1  Apollo12            34.34238                12.78814     Intrepid (LM-6)   
2  Apollo14            41.83363                 7.49125      Antares (LM-8)   
3  Apollo15            75.39910                33.56547      Falcon (LM-10)   
4  Apollo16            92.46262                17.06352       Orion (LM-11)   
5  Apollo17           109.44402                16.98140  Challenger (LM-12)   

   LM Mass (kg)  LM Mass Difference      Command Module (CM)  CM Mass (kg)  \
0         15103                 0.0       Columbia (CSM-107)          5560   
1         15235               132.0  Yankee Clipper (CM-108)          5609   
2         15264                29.0      Kitty Hawk (CM-110)          5758   
3         16430              1166.0        Endeavor (CM-112)          5875   
4         16445                15.0          Casper (CM-113)          5840   
5         16456                11.0         America (CM-114)          5960   

   CM Mass Difference  Total Weight (kg)  Total Difference (kg)  \
0                 0.0              20663                    0.0   
1                49.0              20844                  181.0   
2               149.0              21022                  178.0   
3               117.0              22305                 1283.0   
4               -35.0              22285                  -20.0   
5               120.0              22416                  131.0   

   Crewed Area: Payload  Sample: Crewed Area  Sample: Payload  
0              0.475011             0.001043         0.000495  
1              0.479172             0.001648         0.000789  
2              0.483264             0.001990         0.000962  
3              0.512759             0.003380         0.001733  
4              0.512299             0.004149         0.002126  
5              0.515310             0.004882         0.002516

### Exercise 5: Predict Artemis sample capacity

<font size="3"> **Unit 7 of 11** </font>

<font size="2">Step by step process of **Exercise 5** can be found here: https://learn.microsoft.com/en-us/training/modules/plan-moon-mission-using-python-pandas/7-artemis-capacity </font>

------

<font size="3"> **References/Vocabs:** </font>
<font size="2">
<p>As a reminder, the Artemis program is NASA's second set of missions to land humans on the surface of the Moon. The program will launch in 2025 and will send not only the next pair of humans, but also the first woman to set foot on the Moon. The preparation for this mission is even bigger than focusing on a Moon landing. It will also provide space for a commercial payload on the ship and is the first step along the Moon to Mars program. So, while the Artemis missions will likely bring home additional samples, there are other goals that might affect the amount of capacity that's required to do so.

We don't have all the details about the Artemis mission, but we do know currently that three iterations of the rocket will be cycled through for each mission. Each rocket will have one version meant to sustain a crew and one meant only for cargo. For the purposes of this module, we will focus only on the three rockets meant to house crew, to be more aligned with the Apollo missions. We also know that the expected payload of the Space Launch System (SLS) is expected to grow with each iteration, but that *__the current weight of Orion (the command and lunar modules combined) has one estimated weight today.__*

Again, we will call __the command and lunar modules the__ *__crewed area__*, and we can create a DataFrame with the information we have about the *__three crewed missions__*:</p> </font>

#### Exercise 5.1

In [10]:
import pandas as pd
rock_samples = pd.read_csv('data/rocksamples.csv')

artemis_crewedArea = 26520
artemis_missions = pd.DataFrame({'Mission': ['Artemis 1','Artemis 1b','Artemis 2'],
                                'Total Weight (kg)': [artemis_crewedArea,artemis_crewedArea,artemis_crewedArea],
                                'Payload (kg)': [26988, 37965, 42955]})

artemis_missions

Mission  Total Weight (kg)  Payload (kg)
0   Artemis 1              26520         26988
1  Artemis 1b              26520         37965
2   Artemis 2              26520         42955

#### Exercise 5.2

In [8]:
import pandas as pd
rock_samples = pd.read_csv('data/rocksamples.csv')

rock_samples['Weight (g)'] = rock_samples['Weight (g)'].apply(lambda x : x * 0.001)
rock_samples.rename(columns = {'Weight (g)' : 'Weight (kg)'}, inplace = True)
rock_samples.head()

missions = pd.DataFrame()
missions['Mission'] = rock_samples['Mission'].unique()
missions.head()

sample_total_weight = rock_samples.groupby('Mission')['Weight (kg)'].sum()
missions = pd.merge(missions, sample_total_weight, on = 'Mission')
missions.rename(columns = {'Weight (kg)' : 'Sample Weight (kg)'}, inplace = True)
missions['Weight Difference (kg)'] = missions['Sample Weight (kg)'].diff()
missions['Weight Difference (kg)'] = missions ['Weight Difference (kg)'].fillna(value = 0)

missions['Lunar Module (LM)'] = ['Eagle (LM-5)', 'Intrepid (LM-6)', 'Antares (LM-8)', 'Falcon (LM-10)', 'Orion (LM-11)', 'Challenger (LM-12)']
missions['LM Mass (kg)'] = [15103, 15235, 15264, 16430, 16445, 16456]
missions['LM Mass Difference (kg)'] = missions['LM Mass (kg)'].diff()
missions['LM Mass Difference (kg)'] = missions['LM Mass Difference (kg)'].fillna(value = 0)

missions['Command Module (CM)'] =['Columbia (CSM-107)', 'Yankee Clipper (CM-108)', 'Kitty Hawk (CM-110)', 'Endeavor (CM-112)', 'Casper (CM-113)', 'America (CM-114)']
missions['CM Mass (kg)'] = [5560, 5609, 5758, 5875, 5840, 5960]
missions['CM Mass Difference (kg)'] = missions['CM Mass (kg)'].diff()
missions['CM Mass Difference (kg)'] = missions['CM Mass Difference (kg)'].fillna(value = 0)

missions['Total Weight (kg)'] = missions['LM Mass (kg)'] + missions['CM Mass (kg)']
missions['Total Difference (kg)'] = missions['LM Mass Difference (kg)'] + missions['CM Mass Difference (kg)']

# Sample-to-weight ratio for Apollo missions:
saturnVPayload = 43500 #(kg)
missions['Crewed Area: Payload'] = missions['Total Weight (kg)']/saturnVPayload
missions['Sample: Crewed Area'] = missions['Sample Weight (kg)']/missions['Total Weight (kg)']
missions['Sample: Payload'] = missions['Sample Weight (kg)']/saturnVPayload

missions

# Take the average of all those ratios across all the Apollo missions:
crewedarea_payload_average = missions['Crewed Area: Payload'].mean()
sample_crewedarea_average = missions['Sample: Crewed Area'].mean()
sample_payload_average = missions['Sample: Payload'].mean()

print(crewedarea_payload_average)
print(sample_crewedarea_average)
print(sample_payload_average)

artemis_crewedArea = 26520
artemis_missions = pd.DataFrame({'Mission': ['Artemis 1','Artemis 1b','Artemis 2'],
                                'Total Weight (kg)': [artemis_crewedArea,artemis_crewedArea,artemis_crewedArea],
                                'Payload (kg)': [26988, 37965, 42955]})

# Sample-to-weight ratio for (estimated) Artemis missions:
artemis_missions['Sample Weight from Total Weight (kg)'] = artemis_missions['Total Weight (kg)'] * sample_crewedarea_average
artemis_missions['Sample Weight from Payload (kg)'] = artemis_missions['Payload (kg)'] * sample_payload_average

artemis_missions['Estimated Sample Weight (kg)'] = (artemis_missions['Sample Weight from Total Weight (kg)'] + artemis_missions['Sample Weight from Payload (kg)'])/2

artemis_missions

0.4963026819923371
0.002848764392685611
0.0014369195019157087


Mission  Total Weight (kg)  Payload (kg)  \
0   Artemis 1              26520         26988   
1  Artemis 1b              26520         37965   
2   Artemis 2              26520         42955   

   Sample Weight from Total Weight (kg)  Sample Weight from Payload (kg)  \
0                             75.549232                        38.779584   
1                             75.549232                        54.552649   
2                             75.549232                        61.722877   

   Estimated Sample Weight (kg)  
0                     57.164408  
1                     65.050940  
2                     68.636054

### Exercise 6: Prioritize Moon rock sample gathering based on data

<font size="3"> **Unit 8 of 11** </font>

<font size="2">Step by step process of **Exercise 6** can be found here: https://learn.microsoft.com/en-us/training/modules/plan-moon-mission-using-python-pandas/8-prioritize-needed-rocks </font>

------

<font size="3"> **References/Vocabs:** </font>
<font size="2"><p>
Determining which types of samples to collect from the Moon requires expertise, but we can start to make some assumptions to learn how to clean and manipulate data.

First, we can determine how much remains of each sample that was returned from the Apollo missions by multiplying the weight of the sample that was originally collected by the percentage of remaining pristine sample. </p> </font>

#### Exercise 6.1

In [13]:
import pandas as pd
rock_samples = pd.read_csv('data/rocksamples.csv')

rock_samples['Weight (g)'] = rock_samples['Weight (g)'].apply(lambda x : x * 0.001)
rock_samples.rename(columns = {'Weight (g)' : 'Weight (kg)'}, inplace = True)
rock_samples.head()

rock_samples['Remaining (kg)'] = rock_samples['Weight (kg)'] * (rock_samples['Pristine (%)'] * 0.01)
rock_samples.describe()

ID  Weight (kg)  Pristine (%)  Remaining (kg)
count   2229.000000  2229.000000   2229.000000     2229.000000
mean   52058.432032     0.168253     84.512764        0.138103
std    26207.651471     0.637286     22.057299        0.525954
min    10001.000000     0.000000      0.000000        0.000000
25%    15437.000000     0.003000     80.010000        0.002432
50%    65527.000000     0.010200     92.300000        0.008530
75%    72142.000000     0.093490     98.140000        0.078240
max    79537.000000    11.729000    180.000000       11.169527

#### Exercise 6.2

In [3]:
import pandas as pd
rock_samples = pd.read_csv('data/rocksamples.csv')

rock_samples['Weight (g)'] = rock_samples['Weight (g)'].apply(lambda x : x * 0.001)
rock_samples.rename(columns = {'Weight (g)' : 'Weight (kg)'}, inplace = True)
rock_samples.head()

rock_samples['Remaining (kg)'] = rock_samples['Weight (kg)'] * (rock_samples['Pristine (%)'] * 0.01)
rock_samples.describe()

low_samples = rock_samples.loc[(rock_samples['Weight (kg)'] >= 0.16) & (rock_samples['Pristine (%)'] <= 50)]
low_samples.head()

ID   Mission     Type   Subtype  Weight (kg)  Pristine (%)  \
11  10017  Apollo11   Basalt  Ilmenite        0.973         43.71   
14  10020  Apollo11   Basalt  Ilmenite        0.425         27.88   
15  10021  Apollo11  Breccia  Regolith        0.250         30.21   
29  10045  Apollo11   Basalt   Olivine        0.185         12.13   
37  10057  Apollo11   Basalt  Ilmenite        0.919         35.15   

    Remaining (kg)  
11        0.425298  
14        0.118490  
15        0.075525  
29        0.022441  
37        0.323028

#### Exercise 6.3

In [4]:
import pandas as pd
rock_samples = pd.read_csv('data/rocksamples.csv')

rock_samples['Weight (g)'] = rock_samples['Weight (g)'].apply(lambda x : x * 0.001)
rock_samples.rename(columns = {'Weight (g)' : 'Weight (kg)'}, inplace = True)
rock_samples.head()

rock_samples['Remaining (kg)'] = rock_samples['Weight (kg)'] * (rock_samples['Pristine (%)'] * 0.01)
rock_samples.describe()

low_samples = rock_samples.loc[(rock_samples['Weight (kg)'] >= 0.16) & (rock_samples['Pristine (%)'] <= 50)]
low_samples.info()

<class 'pandas.core.frame.DataFrame'>
Index: 27 entries, 11 to 2183
Data columns (total 7 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   ID              27 non-null     int64  
 1   Mission         27 non-null     object 
 2   Type            27 non-null     object 
 3   Subtype         27 non-null     object 
 4   Weight (kg)     27 non-null     float64
 5   Pristine (%)    27 non-null     float64
 6   Remaining (kg)  27 non-null     float64
dtypes: float64(3), int64(1), object(3)
memory usage: 1.7+ KB


#### Exercise 6.4

In [5]:
import pandas as pd
rock_samples = pd.read_csv('data/rocksamples.csv')

rock_samples['Weight (g)'] = rock_samples['Weight (g)'].apply(lambda x : x * 0.001)
rock_samples.rename(columns = {'Weight (g)' : 'Weight (kg)'}, inplace = True)
rock_samples.head()

rock_samples['Remaining (kg)'] = rock_samples['Weight (kg)'] * (rock_samples['Pristine (%)'] * 0.01)
rock_samples.describe()

low_samples = rock_samples.loc[(rock_samples['Weight (kg)'] >= 0.16) & (rock_samples['Pristine (%)'] <= 50)]
low_samples.head()

low_samples.Type.unique()

array(['Basalt', 'Breccia', 'Soil', 'Core'], dtype=object)

#### Exercise 6.5

In [6]:
import pandas as pd
rock_samples = pd.read_csv('data/rocksamples.csv')

rock_samples['Weight (g)'] = rock_samples['Weight (g)'].apply(lambda x : x * 0.001)
rock_samples.rename(columns = {'Weight (g)' : 'Weight (kg)'}, inplace = True)
rock_samples.head()

rock_samples['Remaining (kg)'] = rock_samples['Weight (kg)'] * (rock_samples['Pristine (%)'] * 0.01)
rock_samples.describe()

low_samples = rock_samples.loc[(rock_samples['Weight (kg)'] >= 0.16) & (rock_samples['Pristine (%)'] <= 50)]
low_samples.head()

rock_samples.Type.unique()

array(['Soil', 'Basalt', 'Core', 'Breccia', 'Special', 'Crustal'],
      dtype=object)

#### Exercise 6.6

In [7]:
import pandas as pd
rock_samples = pd.read_csv('data/rocksamples.csv')

rock_samples['Weight (g)'] = rock_samples['Weight (g)'].apply(lambda x : x * 0.001)
rock_samples.rename(columns = {'Weight (g)' : 'Weight (kg)'}, inplace = True)
rock_samples.head()

rock_samples['Remaining (kg)'] = rock_samples['Weight (kg)'] * (rock_samples['Pristine (%)'] * 0.01)
rock_samples.describe()

low_samples = rock_samples.loc[(rock_samples['Weight (kg)'] >= 0.16) & (rock_samples['Pristine (%)'] <= 50)]
low_samples.head()

low_samples.Type.unique()

low_samples.groupby('Type')['Weight (kg)'].count()

Type
Basalt     14
Breccia     8
Core        1
Soil        4
Name: Weight (kg), dtype: int64

#### Exercise 6.7a

In [11]:
import pandas as pd
rock_samples  = pd.read_csv('data/rocksamples.csv')

rock_samples['Weight (g)'] = rock_samples['Weight (g)'].apply(lambda x : x * 0.001)
rock_samples.rename(columns = {'Weight (g)' : 'Weight (kg)'}, inplace = True)
rock_samples.head()

rock_samples['Remaining (kg)'] = rock_samples['Weight (kg)'] * (rock_samples['Pristine (%)'] * 0.01)
rock_samples.describe()

low_samples = rock_samples.loc[(rock_samples['Weight (kg)'] >= 0.16) & (rock_samples['Pristine (%)'] <=50)]
low_samples.head()

low_samples.Type.unique()
low_samples.groupby('Type')['Weight (kg)'].count()

needed_samples = low_samples[low_samples['Type'].isin(['Basalt', 'Breccia'])]
needed_samples.info()

<class 'pandas.core.frame.DataFrame'>
Index: 22 entries, 11 to 2183
Data columns (total 7 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   ID              22 non-null     int64  
 1   Mission         22 non-null     object 
 2   Type            22 non-null     object 
 3   Subtype         22 non-null     object 
 4   Weight (kg)     22 non-null     float64
 5   Pristine (%)    22 non-null     float64
 6   Remaining (kg)  22 non-null     float64
dtypes: float64(3), int64(1), object(3)
memory usage: 1.4+ KB


#### Exercise 6.7b

In [12]:
import pandas as pd
rock_samples  = pd.read_csv('data/rocksamples.csv')

rock_samples['Weight (g)'] = rock_samples['Weight (g)'].apply(lambda x : x * 0.001)
rock_samples.rename(columns = {'Weight (g)' : 'Weight (kg)'}, inplace = True)
rock_samples.head()

rock_samples['Remaining (kg)'] = rock_samples['Weight (kg)'] * (rock_samples['Pristine (%)'] * 0.01)
rock_samples.describe()

low_samples = rock_samples.loc[(rock_samples['Weight (kg)'] >= 0.16) & (rock_samples['Pristine (%)'] <=50)]
low_samples.head()

low_samples.Type.unique()
low_samples.groupby('Type')['Weight (kg)'].count()

needed_samples = low_samples[low_samples['Type'].isin(['Basalt', 'Breccia'])]
needed_samples.head()

ID   Mission     Type   Subtype  Weight (kg)  Pristine (%)  \
11  10017  Apollo11   Basalt  Ilmenite        0.973         43.71   
14  10020  Apollo11   Basalt  Ilmenite        0.425         27.88   
15  10021  Apollo11  Breccia  Regolith        0.250         30.21   
29  10045  Apollo11   Basalt   Olivine        0.185         12.13   
37  10057  Apollo11   Basalt  Ilmenite        0.919         35.15   

    Remaining (kg)  
11        0.425298  
14        0.118490  
15        0.075525  
29        0.022441  
37        0.323028

### Exercise 7: Develop a recommendation of Moon rock samples to be collected

<font size="3"> **Unit 9 of 11** </font>

<font size="2">Step by step process of **Exercise 7** can be found here: https://learn.microsoft.com/en-us/training/modules/plan-moon-mission-using-python-pandas/9-develop-recommendation </font>

------

#### Exercise 7.1

In [13]:
import pandas as pd
rock_samples = pd.read_csv('data/rocksamples.csv')

rock_samples['Weight (g)'] = rock_samples['Weight (g)'].apply(lambda x : x * 0.001)
rock_samples.rename(columns = {'Weight (g)' : 'Weight (kg)'}, inplace = True)
rock_samples.head()

rock_samples['Remaining (kg)'] = rock_samples['Weight (kg)'] * (rock_samples['Pristine (%)'] * 0.01)
rock_samples.describe()

low_samples = rock_samples.loc[(rock_samples['Weight (kg)'] >= 0.16) & (rock_samples['Pristine (%)'] <= 50)]
low_samples.head()

low_samples.Type.unique()
low_samples.groupby('Type')['Weight (kg)'].count()

needed_samples = low_samples[low_samples['Type'].isin(['Basalt', 'Breccia'])]
needed_samples.head()

needed_samples.groupby('Type')['Weight (kg)'].sum()


Type
Basalt     17.4234
Breccia    10.1185
Name: Weight (kg), dtype: float64

#### Exercise 7.2

In [14]:
import pandas as pd
rock_samples = pd.read_csv('data/rocksamples.csv')

rock_samples['Weight (g)'] = rock_samples['Weight (g)'].apply(lambda x : x * 0.001)
rock_samples.rename(columns = {'Weight (g)' : 'Weight (kg)'}, inplace = True)
rock_samples.head()

rock_samples['Remaining (kg)'] = rock_samples['Weight (kg)'] * (rock_samples['Pristine (%)'] * 0.01)
rock_samples.describe()

low_samples = rock_samples.loc[(rock_samples['Weight (kg)'] >= 0.16) & (rock_samples['Pristine (%)'] <= 50)]
low_samples.head()

low_samples.Type.unique()
low_samples.groupby('Type')['Weight (kg)'].count()

needed_samples = low_samples[low_samples['Type'].isin(['Basalt', 'Breccia'])]
needed_samples.head()

rock_samples.groupby('Type')['Weight (kg)'].sum()


Type
Basalt      93.14077
Breccia    168.88075
Core        19.93587
Crustal      4.74469
Soil        87.58981
Special      0.74410
Name: Weight (kg), dtype: float64

#### Exercise 7.3
<font size="2"><p>
This bit of information really stands out: we didn't have many **Crustal** rocks in the first place.

*We can add* *__Crustal__* *rocks to the set of needed samples:* </p> </font>

In [19]:
import pandas as pd
rock_samples = pd.read_csv('data/rocksamples.csv')

rock_samples['Weight (g)'] = rock_samples['Weight (g)'].apply(lambda x : x * 0.001)
rock_samples.rename(columns = {'Weight (g)' : 'Weight (kg)'}, inplace = True)
rock_samples.head()

rock_samples['Remaining (kg)'] = rock_samples['Weight (kg)'] * (rock_samples['Pristine (%)'] * 0.01)
rock_samples.describe()

low_samples = rock_samples.loc[(rock_samples['Weight (kg)'] >= 0.16) & (rock_samples['Pristine (%)'] <= 50)]
low_samples.head()

low_samples.Type.unique()

needed_samples = low_samples[low_samples['Type'].isin(['Basalt', 'Breccia'])]
needed_samples.head()

needed_samples.groupby('Type')['Weight (kg)'].sum()

needed_samples = pd.concat([needed_samples, rock_samples.loc[rock_samples['Type'] == 'Crustal']])
needed_samples.info()

<class 'pandas.core.frame.DataFrame'>
Index: 68 entries, 11 to 2189
Data columns (total 7 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   ID              68 non-null     int64  
 1   Mission         68 non-null     object 
 2   Type            68 non-null     object 
 3   Subtype         68 non-null     object 
 4   Weight (kg)     68 non-null     float64
 5   Pristine (%)    68 non-null     float64
 6   Remaining (kg)  68 non-null     float64
dtypes: float64(3), int64(1), object(3)
memory usage: 4.2+ KB


#### Exercise 7.4
<font size="3"> **Summary of needed samples** </font>
<font size="2"><p>
The final step is to consolidate everything we know into one table that can be shared with the astronauts. When astronauts are up on the Moon, one way they can identify rocks is by their size. If we can give them an estimated size of each type of rock that might make their collection process easier.<p></font>

In [29]:
import pandas as pd
rock_samples = pd.read_csv('data/rocksamples.csv')

rock_samples['Weight (g)'] = rock_samples['Weight (g)'].apply(lambda x : x * 0.001)
rock_samples.rename(columns = {'Weight (g)' : 'Weight (kg)'}, inplace = True)
rock_samples.head()

rock_samples['Remaining (kg)'] = rock_samples['Weight (kg)'] * (rock_samples['Pristine (%)'] * 0.01)
rock_samples.describe()

low_samples = rock_samples.loc[(rock_samples['Weight (kg)'] >= 0.16) & (rock_samples['Pristine (%)'] <= 50)]
low_samples.head()
low_samples.Type.unique()

needed_samples = low_samples[low_samples['Type'].isin(['Basalt', 'Breccia'])]
needed_samples.groupby('Type')['Weight (kg)'].sum()

needed_samples = pd.concat([needed_samples, rock_samples[rock_samples['Type'] == 'Crustal']])

needed_samples_overview = pd.DataFrame()
needed_samples_overview['Type'] = needed_samples.Type.unique()
needed_samples_overview

Type
0   Basalt
1  Breccia
2  Crustal

#### Exercise 7.5

In [33]:
import pandas as pd
rock_samples = pd.read_csv('data/rocksamples.csv')

rock_samples['Weight (g)'] = rock_samples['Weight (g)'].apply(lambda x : x * 0.001)
rock_samples.rename(columns = {'Weight (g)' : 'Weight (kg)'}, inplace = True)
rock_samples.head()

rock_samples['Remaining (kg)'] = rock_samples['Weight (kg)'] * (rock_samples['Pristine (%)'] * 0.01)
rock_samples.describe()

low_samples = rock_samples.loc[(rock_samples['Weight (kg)'] >= 0.16) & (rock_samples['Pristine (%)'] <= 50)]
low_samples.head()
low_samples.Type.unique()

needed_samples = low_samples[low_samples['Type'].isin(['Basalt', 'Breccia'])]
needed_samples.groupby('Type')['Weight (kg)'].count()
needed_samples.groupby('Type')['Weight (kg)'].sum()

needed_samples = pd.concat([needed_samples, rock_samples[rock_samples['Type'] == 'Crustal']])

needed_samples_overview = pd.DataFrame()
needed_samples_overview['Type'] = needed_samples.Type.unique()

needed_samples_weights = needed_samples.groupby('Type')['Weight (kg)'].sum().reset_index()
needed_samples_overview = pd.merge(needed_samples_overview, needed_samples_weights, on = 'Type')
needed_samples_overview.rename(columns = {'Weight (kg)' : 'Total Weight (kg)'}, inplace = True)
needed_samples_overview

Type  Total Weight (kg)
0   Basalt           17.42340
1  Breccia           10.11850
2  Crustal            4.74469

#### Exercise 7.6

In [5]:
import pandas as pd
rock_samples = pd.read_csv('data/rocksamples.csv')

rock_samples['Weight (g)'] = rock_samples['Weight (g)'].apply(lambda x : x * 0.001)
rock_samples.rename(columns = {'Weight (g)' : 'Weight (kg)'}, inplace = True)
rock_samples.head()

rock_samples['Remaining (kg)'] = rock_samples['Weight (kg)'] * (rock_samples['Pristine (%)'] * 0.01)
rock_samples.describe()

low_samples = rock_samples.loc[(rock_samples['Weight (kg)'] >= 0.16) & (rock_samples['Pristine (%)'] <= 50)]
low_samples.head()
low_samples.Type.unique()

needed_samples = low_samples[low_samples['Type'].isin(['Basalt', 'Breccia'])]
needed_samples.groupby('Type')['Weight (kg)'].count()
needed_samples.groupby('Type')['Weight (kg)'].sum()

needed_samples = pd.concat([needed_samples, rock_samples[rock_samples['Type'] == 'Crustal']])

needed_samples_overview = pd.DataFrame()
needed_samples_overview['Type'] = needed_samples.Type.unique()

needed_samples_weights = needed_samples.groupby('Type')['Weight (kg)'].sum().reset_index()
needed_samples_overview = pd.merge(needed_samples_overview, needed_samples_weights, on = 'Type')
needed_samples_overview.rename(columns = {'Weight (kg)' : 'Total Weight (kg)'}, inplace = True)

needed_samples_ave_weights = needed_samples.groupby('Type')['Weight (kg)'].mean().reset_index()
needed_samples_overview = pd.merge(needed_samples_overview, needed_samples_ave_weights, on = 'Type')
needed_samples_overview.rename(columns = {'Weight (kg)' : 'Average Weight (kg)'}, inplace = True)

needed_samples_overview

Type  Total Weight (kg)  Average Weight (kg)
0   Basalt           17.42340             1.244529
1  Breccia           10.11850             1.264813
2  Crustal            4.74469             0.103145

#### Exercise 7.7
<font size="2"> Crustals are small. They're probably harder to spot, so no wonder we don't have many of them.

We probably want to give the astronauts some indication of how many of each type we want them to collect. So, for the three types we're looking for, we should grab the total number we have of each type and get the remaining percentage of each type of rock.</font>

In [1]:
import pandas as pd
rock_samples = pd.read_csv('data/rocksamples.csv')

rock_samples['Weight (g)'] = rock_samples['Weight (g)'].apply(lambda x : x * 0.001)
rock_samples.rename(columns = {'Weight (g)' : 'Weight (kg)'}, inplace = True)
rock_samples.head()

rock_samples['Remaining (kg)'] = rock_samples['Weight (kg)'] * (rock_samples['Pristine (%)'] * 0.01)
rock_samples.describe()

low_samples = rock_samples.loc[(rock_samples['Weight (kg)'] >= 0.16) & (rock_samples['Pristine (%)'] <= 50)]
low_samples.head()
low_samples.Type.unique()

needed_samples = low_samples[low_samples['Type'].isin(['Basalt', 'Breccia'])]
needed_samples.groupby('Type')['Weight (kg)'].count()
needed_samples.groupby('Type')['Weight (kg)'].sum()

needed_samples = pd.concat([needed_samples, rock_samples[rock_samples['Type'] == 'Crustal']])

needed_samples_overview = pd.DataFrame()
needed_samples_overview['Type'] = needed_samples.Type.unique()

needed_samples_weights = needed_samples.groupby('Type')['Weight (kg)'].sum().reset_index()
needed_samples_overview = pd.merge(needed_samples_overview, needed_samples_weights, on = 'Type')
needed_samples_overview.rename(columns = {'Weight (kg)' : 'Total Weight (kg)'}, inplace = True)

needed_samples_ave_weights = needed_samples.groupby('Type')['Weight (kg)'].mean().reset_index()
needed_samples_overview = pd.merge(needed_samples_overview, needed_samples_ave_weights, on = 'Type')
needed_samples_overview.rename(columns = {'Weight (kg)' : 'Average Weight (kg)'}, inplace = True)

total_rock_count = rock_samples.groupby('Type')['ID'].count().reset_index()
needed_samples_overview = pd.merge(needed_samples_overview, total_rock_count, on = 'Type')
needed_samples_overview.rename(columns = {'ID' : 'Number of Samples'}, inplace = True)

total_rocks = needed_samples_overview['Number of Samples'].sum()
needed_samples_overview['Percentage of Rocks'] = needed_samples_overview['Number of Samples']/total_rocks
needed_samples_overview

Type  Total Weight (kg)  Average Weight (kg)  Number of Samples  \
0   Basalt           17.42340             1.244529                351   
1  Breccia           10.11850             1.264813                959   
2  Crustal            4.74469             0.103145                 46   

   Percentage of Rocks  
0             0.258850  
1             0.707227  
2             0.033923

#### Exercise 7.8

<font size="2"> And finally, to tie it all back into a recommendation to the Artemis program, we can determine the average weight of samples we estimated in the preceding unit. </font>

In [21]:
import pandas as pd
rock_samples = pd.read_csv('data/rocksamples.csv')

rock_samples['Weight (g)'] = rock_samples['Weight (g)'].apply(lambda x : x * 0.001)
rock_samples.rename(columns = {'Weight (g)' : 'Weight (kg)'}, inplace = True)
rock_samples.head()

rock_samples['Remaining (kg)'] = rock_samples['Weight (kg)'] * (rock_samples['Pristine (%)'] * 0.01)
rock_samples.describe()

low_samples = rock_samples.loc[(rock_samples['Weight (kg)'] >= 0.16) & (rock_samples['Pristine (%)'] <= 50)]
low_samples.head()
low_samples.Type.unique()

needed_samples = low_samples[low_samples['Type'].isin(['Basalt', 'Breccia'])]
needed_samples.groupby('Type')['Weight (kg)'].count()
needed_samples.groupby('Type')['Weight (kg)'].sum()

needed_samples = pd.concat([needed_samples, rock_samples[rock_samples['Type'] == 'Crustal']])

needed_samples_overview = pd.DataFrame()
needed_samples_overview['Type'] = needed_samples.Type.unique()

needed_samples_weights = needed_samples.groupby('Type')['Weight (kg)'].sum().reset_index()
needed_samples_overview = pd.merge(needed_samples_overview, needed_samples_weights, on = 'Type')
needed_samples_overview.rename(columns = {'Weight (kg)' : 'Total Weight (kg)'}, inplace = True)

needed_samples_ave_weights = needed_samples.groupby('Type')['Weight (kg)'].mean().reset_index()
needed_samples_overview = pd.merge(needed_samples_overview, needed_samples_ave_weights, on = 'Type')
needed_samples_overview.rename(columns = {'Weight (kg)' : 'Average Weight (kg)'}, inplace = True)

total_rock_count = rock_samples.groupby('Type')['ID'].count().reset_index()
needed_samples_overview = pd.merge(needed_samples_overview, total_rock_count, on = 'Type')
needed_samples_overview.rename(columns = {'ID' : 'Number of Samples'}, inplace = True)

total_rocks = needed_samples_overview['Number of Samples'].sum()
needed_samples_overview['Percentage of Rocks (%)'] = needed_samples_overview['Number of Samples']/total_rocks
needed_samples_overview['Percentage of Rocks (%)'] = needed_samples_overview['Percentage of Rocks (%)'].apply(lambda x : x * 100)

missions = pd.DataFrame()
missions['Mission'] = rock_samples['Mission'].unique()
missions.head()

sample_total_weight = rock_samples.groupby('Mission')['Weight (kg)'].sum()
missions = pd.merge(missions, sample_total_weight, on = 'Mission')
missions.rename(columns = {'Weight (kg)' : 'Sample Weight (kg)'}, inplace = True)
missions['Weight Difference (kg)'] = missions['Sample Weight (kg)'].diff()
missions['Weight Difference (kg)'] = missions ['Weight Difference (kg)'].fillna(value = 0)

missions['Lunar Module (LM)'] = ['Eagle (LM-5)', 'Intrepid (LM-6)', 'Antares (LM-8)', 'Falcon (LM-10)', 'Orion (LM-11)', 'Challenger (LM-12)']
missions['LM Mass (kg)'] = [15103, 15235, 15264, 16430, 16445, 16456]
missions['LM Mass Difference (kg)'] = missions['LM Mass (kg)'].diff()
missions['LM Mass Difference (kg)'] = missions['LM Mass Difference (kg)'].fillna(value = 0)

missions['Command Module (CM)'] =['Columbia (CSM-107)', 'Yankee Clipper (CM-108)', 'Kitty Hawk (CM-110)', 'Endeavor (CM-112)', 'Casper (CM-113)', 'America (CM-114)']
missions['CM Mass (kg)'] = [5560, 5609, 5758, 5875, 5840, 5960]
missions['CM Mass Difference (kg)'] = missions['CM Mass (kg)'].diff()
missions['CM Mass Difference (kg)'] = missions['CM Mass Difference (kg)'].fillna(value = 0)

missions['Total Weight (kg)'] = missions['LM Mass (kg)'] + missions['CM Mass (kg)']
missions['Total Difference (kg)'] = missions['LM Mass Difference (kg)'] + missions['CM Mass Difference (kg)']

# Sample-to-weight ratio for Apollo missions:
saturnVPayload = 43500 #(kg)
missions['Crewed Area: Payload'] = missions['Total Weight (kg)']/saturnVPayload
missions['Sample: Crewed Area'] = missions['Sample Weight (kg)']/missions['Total Weight (kg)']
missions['Sample: Payload'] = missions['Sample Weight (kg)']/saturnVPayload

missions

# Take the average of all those ratios across all the Apollo missions:
crewedarea_payload_average = missions['Crewed Area: Payload'].mean()
sample_crewedarea_average = missions['Sample: Crewed Area'].mean()
sample_payload_average = missions['Sample: Payload'].mean()

print(crewedarea_payload_average)
print(sample_crewedarea_average)
print(sample_payload_average)

artemis_crewedArea = 26520
artemis_missions = pd.DataFrame({'Mission': ['Artemis 1','Artemis 1b','Artemis 2'],
                                'Total Weight (kg)': [artemis_crewedArea,artemis_crewedArea,artemis_crewedArea],
                                'Payload (kg)': [26988, 37965, 42955]})

# Sample-to-weight ratio for (estimated) Artemis missions:
artemis_missions['Sample Weight from Total Weight (kg)'] = artemis_missions['Total Weight (kg)'] * sample_crewedarea_average
artemis_missions['Sample Weight from Payload (kg)'] = artemis_missions['Payload (kg)'] * sample_payload_average

artemis_missions['Estimated Sample Weight (kg)'] = (artemis_missions['Sample Weight from Total Weight (kg)'] + artemis_missions['Sample Weight from Payload (kg)'])/2

artemis_missions

artemis_ave_weight = artemis_missions['Estimated Sample Weight (kg)'].mean()
artemis_ave_weight

needed_samples_overview

0.4963026819923371
0.002848764392685611
0.0014369195019157087


Type  Total Weight (kg)  Average Weight (kg)  Number of Samples  \
0   Basalt           17.42340             1.244529                351   
1  Breccia           10.11850             1.264813                959   
2  Crustal            4.74469             0.103145                 46   

   Percentage of Rocks (%)  
0                25.884956  
1                70.722714  
2                 3.392330

#### Exercise 7.9

<font size="2"> We can use this number to determine how many of each rock we want the astronauts to aim to collect: </font>

In [24]:
import pandas as pd
rock_samples = pd.read_csv('data/rocksamples.csv')

rock_samples['Weight (g)'] = rock_samples['Weight (g)'].apply(lambda x : x * 0.001)
rock_samples.rename(columns = {'Weight (g)' : 'Weight (kg)'}, inplace = True)
rock_samples.head()

rock_samples['Remaining (kg)'] = rock_samples['Weight (kg)'] * (rock_samples['Pristine (%)'] * 0.01)
rock_samples.describe()

low_samples = rock_samples.loc[(rock_samples['Weight (kg)'] >= 0.16) & (rock_samples['Pristine (%)'] <= 50)]
low_samples.head()
low_samples.Type.unique()

needed_samples = low_samples[low_samples['Type'].isin(['Basalt', 'Breccia'])]
needed_samples.groupby('Type')['Weight (kg)'].count()
needed_samples.groupby('Type')['Weight (kg)'].sum()

needed_samples = pd.concat([needed_samples, rock_samples[rock_samples['Type'] == 'Crustal']])

needed_samples_overview = pd.DataFrame()
needed_samples_overview['Type'] = needed_samples.Type.unique()

needed_samples_weights = needed_samples.groupby('Type')['Weight (kg)'].sum().reset_index()
needed_samples_overview = pd.merge(needed_samples_overview, needed_samples_weights, on = 'Type')
needed_samples_overview.rename(columns = {'Weight (kg)' : 'Total Weight (kg)'}, inplace = True)

needed_samples_ave_weights = needed_samples.groupby('Type')['Weight (kg)'].mean().reset_index()
needed_samples_overview = pd.merge(needed_samples_overview, needed_samples_ave_weights, on = 'Type')
needed_samples_overview.rename(columns = {'Weight (kg)' : 'Average Weight (kg)'}, inplace = True)

total_rock_count = rock_samples.groupby('Type')['ID'].count().reset_index()
needed_samples_overview = pd.merge(needed_samples_overview, total_rock_count, on = 'Type')
needed_samples_overview.rename(columns = {'ID' : 'Number of Samples'}, inplace = True)

total_rocks = needed_samples_overview['Number of Samples'].sum()
needed_samples_overview['Percentage of Rocks (%)'] = needed_samples_overview['Number of Samples']/total_rocks
needed_samples_overview['Percentage of Rocks (%)'] = needed_samples_overview['Percentage of Rocks (%)'].apply(lambda x : x * 100)

missions = pd.DataFrame()
missions['Mission'] = rock_samples['Mission'].unique()
missions.head()

sample_total_weight = rock_samples.groupby('Mission')['Weight (kg)'].sum()
missions = pd.merge(missions, sample_total_weight, on = 'Mission')
missions.rename(columns = {'Weight (kg)' : 'Sample Weight (kg)'}, inplace = True)
missions['Weight Difference (kg)'] = missions['Sample Weight (kg)'].diff()
missions['Weight Difference (kg)'] = missions ['Weight Difference (kg)'].fillna(value = 0)

missions['Lunar Module (LM)'] = ['Eagle (LM-5)', 'Intrepid (LM-6)', 'Antares (LM-8)', 'Falcon (LM-10)', 'Orion (LM-11)', 'Challenger (LM-12)']
missions['LM Mass (kg)'] = [15103, 15235, 15264, 16430, 16445, 16456]
missions['LM Mass Difference (kg)'] = missions['LM Mass (kg)'].diff()
missions['LM Mass Difference (kg)'] = missions['LM Mass Difference (kg)'].fillna(value = 0)

missions['Command Module (CM)'] =['Columbia (CSM-107)', 'Yankee Clipper (CM-108)', 'Kitty Hawk (CM-110)', 'Endeavor (CM-112)', 'Casper (CM-113)', 'America (CM-114)']
missions['CM Mass (kg)'] = [5560, 5609, 5758, 5875, 5840, 5960]
missions['CM Mass Difference (kg)'] = missions['CM Mass (kg)'].diff()
missions['CM Mass Difference (kg)'] = missions['CM Mass Difference (kg)'].fillna(value = 0)

missions['Total Weight (kg)'] = missions['LM Mass (kg)'] + missions['CM Mass (kg)']
missions['Total Difference (kg)'] = missions['LM Mass Difference (kg)'] + missions['CM Mass Difference (kg)']

# Sample-to-weight ratio for Apollo missions:
saturnVPayload = 43500 #(kg)
missions['Crewed Area: Payload'] = missions['Total Weight (kg)']/saturnVPayload
missions['Sample: Crewed Area'] = missions['Sample Weight (kg)']/missions['Total Weight (kg)']
missions['Sample: Payload'] = missions['Sample Weight (kg)']/saturnVPayload

missions

# Take the average of all those ratios across all the Apollo missions:
crewedarea_payload_average = missions['Crewed Area: Payload'].mean()
sample_crewedarea_average = missions['Sample: Crewed Area'].mean()
sample_payload_average = missions['Sample: Payload'].mean()

print(crewedarea_payload_average)
print(sample_crewedarea_average)
print(sample_payload_average)

artemis_crewedArea = 26520
artemis_missions = pd.DataFrame({'Mission': ['Artemis 1','Artemis 1b','Artemis 2'],
                                'Total Weight (kg)': [artemis_crewedArea,artemis_crewedArea,artemis_crewedArea],
                                'Payload (kg)': [26988, 37965, 42955]})

# Sample-to-weight ratio for (estimated) Artemis missions:
artemis_missions['Sample Weight from Total Weight (kg)'] = artemis_missions['Total Weight (kg)'] * sample_crewedarea_average
artemis_missions['Sample Weight from Payload (kg)'] = artemis_missions['Payload (kg)'] * sample_payload_average

artemis_missions['Estimated Sample Weight (kg)'] = (artemis_missions['Sample Weight from Total Weight (kg)'] + artemis_missions['Sample Weight from Payload (kg)'])/2

artemis_missions

artemis_ave_weight = artemis_missions['Estimated Sample Weight (kg)'].mean()
artemis_ave_weight

needed_samples_overview['Weight to Collect (kg)'] = needed_samples_overview['Percentage of Rocks (%)'] * artemis_ave_weight
needed_samples_overview['Weight to Collect (kg)'] = needed_samples_overview['Weight to Collect (kg)'].apply(lambda x : x * 0.01)

needed_samples_overview['Rocks to Collect'] = needed_samples_overview['Weight to Collect (kg)']/needed_samples_overview['Average Weight (kg)']

needed_samples_overview

0.4963026819923371
0.002848764392685611
0.0014369195019157087


Type  Total Weight (kg)  Average Weight (kg)  Number of Samples  \
0   Basalt           17.42340             1.244529                351   
1  Breccia           10.11850             1.264813                959   
2  Crustal            4.74469             0.103145                 46   

   Percentage of Rocks (%)  Weight to Collect (kg)  Rocks to Collect  
0                25.884956               16.467267         13.231731  
1                70.722714               44.991764         35.571884  
2                 3.392330                2.158103         20.922917

<font size="2"> So, we might tell the Artemis astronauts to try to collect *__13 Basalt rocks, 35 Breccia rocks, and 20 Crustal rocks.__* </font> 

### Conclusion: Knowledge check

<font size="3"> **Unit 10 of 11** </font>

<font size="2">Step by step process of **Conclusion: knowledge check** can be found here: https://learn.microsoft.com/en-us/training/modules/plan-moon-mission-using-python-pandas/10-knowledge-check </font>

------

<font size="3"> **Question 1:** </font>
<font size="2">

*__"If you have a pandas DataFrame called samples, what would the output be for samples.head()?__*

<p>

- The top row of the DataFrame
- The top five rows of the DataFrame [**CORRECT:** _if you add a number (for example, head(x)) it would print the top x rows._]
- 10 random rows from the DataFrame
- The column names of the DataFrame 
</p> </font>

<font size="3"> **Question 2:** </font>
<font size="2">

*__"In pandas, DataFrame and series are two names for the same kind of data type."__*

<p>

- True
- False [**CORRECT:** _DataFrames and series are different. DataFrames represent two-dimensional data, like in a table or an Excel worksheet, and series represent one-dimensional data, like a list._]
</p> </font>

<font size="3"> **Question 3:** </font>
<font size="2">

*__"What is the total weight of the samples that were brought back by the six Apollo missions that landed on the Moon?"__*

<p>

- Over 825 lb, or 375 kg [**CORRECT:** _Across 2,229 samples, over 800 lb (375 kg) of sample was returned to Earth._]
- Over 63 lb, or 28 kg
- Over 136 lb, or 62 kg
- Over 490 lb, or 92 kg
</p> </font>

### Summary

<font size="3"> **Unit 11 of 11** </font>

<font size="2">**Module Summary** can be found here: https://learn.microsoft.com/en-us/training/modules/plan-moon-mission-using-python-pandas/11-summary </font>

------

<font size="2">

**In this module, you have:**

1. Gathered information about samples brought back from the Moon via the Apollo missions.<p>
2. Acquired data about the types of spacecraft and rockets used for the Apollo and upcoming Artemis missions. <p>
3. Compiled DataFrames, or tables, of that data, which tells stories and provides insights. <p>
4. Created a prediction of how much sample weight could be returned from the Artemis missions. <p>
5. Made a recommendation for the amount and types of rocks the Artemis astronauts should focus their efforts on, based on the rocks that are currently being used for research here on Earth.

</font>